pkg/xds/generator/prometheus_endpoint_generator.go

```diff
@@ -0,0 +1,84 @@
+package generator
+
+import (
+	"net"
+
+	"github.com/golang/protobuf/proto"
+
+	mesh_proto "github.com/Kong/kuma/api/mesh/v1alpha1"
+	core_xds "github.com/Kong/kuma/pkg/core/xds"
+	xds_context "github.com/Kong/kuma/pkg/xds/context"
+
+	"github.com/Kong/kuma/pkg/xds/envoy"
+)
+
+// PrometheusEndpointGenerator generates an inbound Envoy listener
+// that forwards HTTP requests into the `/stats/prometheus`
+// endpoint of the Envoy Admin API.
+//
+// When generating such a listener, it's important not to overshadow
+// a port that is already in use by the application or other Envoy listeners.
+// In the latter case we prefer not generate Prometheus endpoint at all
+// rather than introduce undeterministic behaviour.
+type PrometheusEndpointGenerator struct {
+}
+
+func (g PrometheusEndpointGenerator) Generate(ctx xds_context.Context, proxy *core_xds.Proxy) ([]*Resource, error) {
+	meshLevelSettings := ctx.Mesh.Resource.Spec.Metrics.GetPrometheus()
+	if meshLevelSettings == nil {
+		// Prometheus metrics must be enabled Mesh-wide for Prometheus endpoint to be generated.
+		return nil, nil
+	}
+	if proxy.Metadata.GetAdminPort() == 0 {
+		// It's not possible to export Prometheus metrics if Envoy Admin API has not been enabled on that dataplane.
```

Comment:
I'd also argue that this is most likely a configuration error. Should we log it?

Reply:
Why error ? We do allow users to turn Envoy Admin API off.

I agree that we should communicate this feedback to a user, but I doubt that logging is an effective way of doing that.

I will create a proposal to include configuration errors into `DataplaneInsights` to make them visible via any supported interface, i.e. `kumactl`, `Kuma UI`, REST API

Reply:
> Why error ? We do allow users to turn Envoy Admin API off.

We do, but I'm not sure the user will know that turning Admin API will result in the lack of metrics. Anyway, let's discuss it further in the issue you've created.